aeppl/logpdf.py

```diff
@@ -0,0 +1,455 @@
+def lognormal_logpdf(op, value, *inputs, **kwargs):
+    mu, sigma = inputs[3:]
+    res = (
+        -0.5 * at.pow((at.log(value) - mu) / sigma, 2)
+        - 0.5 * at.log(2.0 * np.pi)
+        - at.log(sigma)
+        - at.log(value)
+    )
+    res = at.switch(at.gt(value, 0.0), res, -np.inf)
+    res = Assert("sigma > 0")(res, at.all(at.gt(sigma, 0)))
+    return res
+
+
+@_logpdf.register(arb.ParetoRV)
+def pareto_logpdf(op, value, *inputs, **kwargs):
+    alpha, m = inputs[3:]
+    res = at.log(alpha) + xlogy0(alpha, m) - xlogy0(alpha + 1.0, value)
+    res = at.switch(at.ge(value, m), res, -np.inf)
+    res = Assert("alpha > 0, m > 0")(
+        res, at.all(at.gt(alpha, 0.0)), at.all(at.gt(m, 0.0))
+    )
+    return res
+
+
+@_logpdf.register(arb.CauchyRV)
+def cauchy_logpdf(op, value, *inputs, **kwargs):
+    alpha, beta = inputs[3:]
+    res = -at.log(np.pi) - at.log(beta) - at.log1p(at.pow((value - alpha) / beta, 2))
+    res = Assert("beta > 0")(res, at.all(at.gt(beta, 0.0)))
+    return res
+
+
+@_logpdf.register(arb.HalfCauchyRV)
+def halfcauchy_logpdf(op, value, *inputs, **kwargs):
+    res = at.log(2) + cauchy_logpdf(op, value, *inputs, **kwargs)
+    loc, _ = inputs[3:]
+    res = at.switch(at.ge(value, loc), res, -np.inf)
+    return res
+
+
+@_logpdf.register(arb.GammaRV)
+def gamma_logpdf(op, value, *inputs, **kwargs):
+    alpha, inv_beta = inputs[3:]
+    beta = at.reciprocal(inv_beta)
+    res = (
+        -at.gammaln(alpha)
+        + xlogy0(alpha, beta)
+        - beta * value
+        + xlogy0(alpha - 1, value)
+    )
+    res = at.switch(at.ge(value, 0.0), res, -np.inf)
+    res = Assert("alpha > 0, beta > 0")(
+        res, at.all(at.gt(alpha, 0.0)), at.all(at.gt(beta, 0.0))
+    )
+    return res
+
+
+@_logpdf.register(arb.InvGammaRV)
+def invgamma_logpdf(op, value, *inputs, **kwargs):
+    alpha, beta = inputs[3:]
+    res = -(alpha + 1) * np.log(value) - at.gammaln(alpha) - 1.0 / value
+    res = (
+        -at.gammaln(alpha)
+        + xlogy0(alpha, beta)
+        - beta / value
+        + xlogy0(-alpha - 1, value)
+    )
+    res = at.switch(at.ge(value, 0.0), res, -np.inf)
+    res = Assert("alpha > 0, beta > 0")(
+        res, at.all(at.gt(alpha, 0.0)), at.all(at.gt(beta, 0.0))
+    )
+    return res
+
+
+@_logpdf.register(arb.WeibullRV)
+def weibull_logpdf(op, value, *inputs, **kwargs):
+    alpha, beta = inputs[3:]
+    res = (
+        at.log(alpha)
+        - at.log(beta)
+        + (alpha - 1.0) * at.log(value / beta)
+        - at.pow(value / beta, alpha)
+    )
+    res = at.switch(at.ge(value, 0.0), res, -np.inf)
+    res = Assert("alpha > 0, beta > 0")(
+        res, at.all(at.gt(alpha, 0.0)), at.all(at.gt(beta, 0.0))
+    )
+    return res
+
+
+@_logpdf.register(arb.VonMisesRV)
+def vonmises_logpdf(op, value, *inputs, **kwargs):
+    mu, kappa = inputs[3:]
+    res = kappa * at.cos(mu - value) - at.log(2 * np.pi) - at.log(at.i0(kappa))
+    # This doesn't match `scipy.stats.vonmises.logpdf`:
```

Comment:
I am not sure about the logoed, but in general thenscipy vonmises is really poorly implemented, special when it comes to the circular limits and mu translations

Reply:
Let's create a separate issue for this.